### PR TITLE
refactor: Use JSON request snapshots when possible

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

Fallbacks for HTTP body snapshots format: `GraphQL` -> `JSON` -> `plain string`

## Context:

Ref #8386

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: test/graphql-snapshot.ts [hunk 1]
```diff
@@ -177,8 +177,14 @@ function simplifyGraphqlTree(tree: DocumentNode): GraphqlSnapshot {
   return result;
 }
 
-export function makeGraphqlSnapshot(requestBodyStr: string): GraphqlSnapshot {
-  const requestBody = JSON.parse(requestBodyStr);
+interface GraphqlSnapshotInput {
```
Comment (viceice) [inline test/graphql-snapshot.ts:180]:
```suggestion
export interface GraphqlSnapshotInput {
```

It's used in public function declaration

File: test/http-mock.ts [hunk 2]
```diff
@@ -61,10 +61,12 @@ export function scope(basePath: BasePath, options?: nock.Options): nock.Scope {
     const body = req.requestBodyBuffers?.[0]?.toString();
 
     if (body) {
+      let jsonBody: any;
       try {
-        result.graphql = makeGraphqlSnapshot(body);
+        jsonBody = JSON.parse(body);
```
Comment (viceice) [inline test/http-mock.ts:66]: We should check contentType for `application/json` so we don't have too many json parse failues?

**Comment by rarkins** [comment]
Needs de-conflicting

File: test/graphql-snapshot.ts [hunk 3]
```diff
@@ -177,13 +177,23 @@ function simplifyGraphqlTree(tree: DocumentNode): GraphqlSnapshot {
...
+  query: string;
+  variables: Record<string, string>;
+}
+
+export function makeGraphqlSnapshot(
+  requestBody: GraphqlSnapshotInput
+): GraphqlSnapshot | null {
+  try {
+    const { query: queryStr, variables } = requestBody;
+    const queryRawTree = parse(queryStr, { noLocation: true });
```
Comment (viceice) [inline test/graphql-snapshot.ts:190]:
```suggestion
    if (!queryStr) {
        return null;
    }
    const queryRawTree = parse(queryStr, { noLocation: true });
```

Do not try to parse undefined ?

File: test/http-mock.ts [hunk 4]
```diff
@@ -58,12 +58,14 @@ export function scope(basePath: BasePath, options?: nock.Options): nock.Scope {
...
-        result.graphql = makeGraphqlSnapshot(body);
-      } catch (ex) {
+    if (requestBody && headers['content-type'] === 'application/json') {
+      const body = JSON.parse(requestBody);
+      const graphql = makeGraphqlSnapshot(body);
+      if (graphql) {
+        result.graphql = graphql;
+      } else {
         result.body = body;
       }
```
Comment (viceice) [inline test/http-mock.ts:70]: I think a try catch makes sense here to be safe?  or should we simply let it crash with unhandled exception if body isn't valid json? 🤔

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**